### ISSUE TITLE
Contradictory error when trying to inherit primitive

### ISSUE THREAD

**Issue by ozra** [crystal-lang/crystal#959]
I guess there's alot of magic involved with the primitives:

```
Error in ./src/test.cr:47: Int64 is not a class, it's a struct

struct TimeStamp < Int64
```

On a sidenote - how would I do that? (Make a distinct type that acts like Int64, but with additional methods etc.)

**Comment by waterlink** [comment]
You can try something along these lines: http://play.crystal-lang.org/#/r/6uo

**Comment by ozra** [comment]
@waterlink - Thanks, I'll try it out!

**Comment by bcardiff** [comment]
@ozra structs does not add overhead to wrapped types. So you can do something like http://play.crystal-lang.org/#/r/6ur . With method missing and macros you can forward lots of Int functionality.

Is this what you want to do?

The compiler tracks the type of the expression/variable and applies the right functions the values, despite of being just an Int in memory.

**Comment by kostya** [comment]
http://play.crystal-lang.org/#/r/6us

**Comment by ozra** [comment]
Thanks for all the tips. I initially implemented it as a struct with "value" in it, similar to some of above examples, just wanted to see if I could get all the Int64 methods "for free" by inheriting.
The struct + method missing delegation via macros definitely is the road!

I'll leave this issue open, since some other error message probably should be put in place.

**Comment by asterite** [comment]
I think in the case of Timestamp I would use a struct to wrap the int and then define the methods I want. I see Timestamp as similar to Time, and we do have a Time which is just a wrapper to a primitive but doesn't use method_missing. For example I never needed to multiply a time, or to shift times, or doing bitwise operations like & and |, or doing `Timespan.new(100).times { ... }`.

In short: I wouldn't use method_missing or complete delegation in this case. In fact, I would only use it in very few cases. In the compiler we are using it, in [a wrapper to an LLVM::Builder](https://github.com/manastech/crystal/blob/master/src/compiler/crystal/codegen/crystal_llvm_builder.cr), but here the use case seems better, because we are intercepting most methods and adding extra funcionality, but this wrapped builder is still a builder, while I would say that a timestamp, although it looks like an integer from the outside, would only be used for a few math operations (plus and minus, and disallow floats, for example).

But that's my point of view :-)

Another thing that we have to think is to probably mark these primitive values as final, so that they can't be inherited. And then think if we want a final keyword. I personally would like to have `final`, I don't want String to be inherited, for example.

**Comment by ozra** [comment]
@asterite - This is how I did end up implementing it :-) For TimeDuration I also implemented '*', for TimeStamp +/-/</> was enough. And then weeks,days,hours,minutes,seconds,millis methods on Int64/Int32. Worked out fine :-) Implemented for "compatibility" of code I'm converting, so I'll probably rewrite to use Crystal stdlib Time instead later on.

`final` sounds like a good idea, so that there's a "formal reason" for why they're not inheritable.
